src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php

```diff
@@ -164,4 +165,30 @@ public function compile($resolveEnvPlaceholders = false)
     {
         throw new LogicException(sprintf('Cannot compile the container in extension "%s".', $this->extensionClass));
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolveEnvPlaceholders($value, $format = null, array &$usedEnvs = null)
+    {
+        if (true !== $format || !\is_string($value)) {
+            return parent::resolveEnvPlaceholders($value, $format, $usedEnvs);
+        }
+
+        $bag = $this->getParameterBag();
+        $value = $bag->resolveValue($value);
+
+        foreach ($bag->getEnvPlaceholders() as $env => $placeholders) {
+            if (false === strpos($env, ':')) {
+                continue;
+            }
+            foreach ($placeholders as $placeholder) {
+                if (false !== stripos($value, $placeholder)) {
+                    throw new RuntimeException(sprintf('Prefixed "env(%s)" parameter is incompatible with load-time resolution in "%s". The inlining logic should be moved to a compiler pass, or a non-prefixed parameter should be used instead.', $env, $this->extensionClass));
```

Comment:
I would not use the term prefix that does not really tells me what you are talking about (especially as `int` `env(int:foo)` is not even a prefix but inside `env()`). I would use `cast` instead of `prefix` here.

What about: 

`Using a cast in "env(%s)" is incompatible with resolution at compile time in "%s". The inlining logic should be moved to a compiler pass, or a parameter without casting should be used instead.`